### ISSUE TITLE
Support for "NAck"

### ISSUE THREAD

**Issue by bruth** [nats-io/stan.go#167]
This applies when manual Ack-ing is in use to provide a way to tell the server to fail the acknowledgement and retry again.

This has the advantage of not having to wait for the `AckWait` timeout (which is often a upper bound guess anyway). And it could also improve the programmatic model if Ack/NAck are exclusive, similar to a SQL commit/rollback. You can defer a `NAck()` and `Ack()` when successful or vice versa.

```go
sc.Subscribe("foo", func(msg *stan.Msg) {
  defer msg.NAck()

  // Conditions to Ack..
}, stan.SetManualAckMode())
```

**Comment by thedodd** [comment]
This would be an awesome feature to have.

**Comment by thedodd** [comment]
Any of the core maintainers for this system have any feedback or thoughts on this? Is it something that is being considered for future development?

**Comment by kozlovic** [comment]
The problem I have with NAck and NATS Streaming is due to the relationship between the NATS Streaming Server and its clients (see https://github.com/nats-io/nats-streaming-server#relation-to-nats).

What that means is that if the server sends a message to a client, it won't know if that message is received or not. The client may not even know that it did not get this message. With AckWait, message will be redelivered and once connectivity is resumed, client will ultimately receive this message.

That being said, server sends heartbeats to clients which could help detect that the client is gone, and in some of the client libraries, we also have introduced PINGs to the server with a shorter default frequency. 
If we introduce the notion of NAck, we would need to make that a subscription option (to prevent the default behavior of the server of resending messages passed the AckWait) and obviously the NAck protocol itself.
But again, due to lack of direct TCP connection between the server and client, I am afraid that this mode would cause more issues than it tries to solve. For instance, if the server sends say 10 messages to a client (due to MaxInflight) then crashes, then client declares the connection lost. The server restarts, and the application recreates the connection and sub: since the server "auto ack" messages as there are sent, it would start delivering messages past the last message that was sent before the crash. Unless the application takes responsibility from recreating the subscription at a given sequence number (the one that it last received), messages will be missing.

**Comment by thedodd** [comment]
@kozlovic thanks for the feedback. Yea, the idea I had in mind was to leverage the heartbeat system, as it is already used to remove unhealthy clients and such.

Agreed, the notion of "resend after timeout" & "resend after nack" are a bit exclusive. However, in your least paragraph, I'm not sure I understand what you mean. Why would the server "auto ack" messages as they are sent? From what I recall, it is the responsibility of the client to ack, which can either be manual or auto. Not sure why the server would be doing that.

All in all, perhaps only a simple update to the streaming protocol would be needed:
- if timeout is set to 0, then the clients must ack/nack messages. Resend after timeout will be disabled for the target stream.
- clients could still use auto ack.
- more complex workflows will be able to manually ack once successful processing is done, and will nack if errors are encountered, which will cause re-delivery.
- if the client is removed because it has failed heartbeats &c, then any outstanding messages which were sent to that client will be re-delivered to others. This is pretty standard in AMQP land.

Thoughts?

**Comment by kozlovic** [comment]
I think I missed the point of the original poster that the option of a NAck() does prevent the use of Ack(), that is, default behavior would be the same as it is today: if server does not receive an ACK, it would redeliver a message after AckWait, but the difference is that if a NAck() is received, then server would redeliver THAT message immediately (and reset its next redelivery timestamp).
What the OP mentioned is that if server receives a NAck after an Ack (for a given message), the NAck would be ignored. 
This still requires a new protocol message (the NAck) but seem much more doable. Will noodle on this a little bit, but seem like something that we could definitively add.

**Comment by molon** [comment]
Support for `NAck(customAckWait)` and returns `redelivery_count` in `MsgProto` would be excellent.

**Comment by thedodd** [comment]
Hey @kozlovic, just wanted to ping you and see if you had any additional thoughts or updates on this.

**Comment by kozlovic** [comment]
We have been discussing this internally, and we think that there may be a different approach. If the issue is that an application is not able to process a message, it is likely that it will be the same for the next. So instead of having a NAck(), it would be worth closing the subscription instead, and wait for the reason that causes the inability to process a message to get resolved. When that's done, restart the subscription. For (durable) queue members, this PR: https://github.com/nats-io/nats-streaming-server/pull/729 will help because the messages assigned to the queue member being closed will immediately be redelivered to other members of the same group.

**Comment by PatrLind** [comment]
I just hit this issue, and I thought I could add my use case in case it is useful for you.

I have a queue of messages that should be processed sequentially. The processing time can be quite lengthy (from a few seconds up to an hour or so). When the processing is done I Ack the message to the STAN server to indicate that it is done. If there is a temporary processing error (resource problem, etc) the message could be Nack:ed so it can be retried at a later time.
Also, unfortunately as it is now, I have to set a very long AckWait time, otherwise the message would be retied over and over again before the processing was done.

Another approach that could work is if there were a way to tell the server that I am still working on a message. The server could then reset the AckWait time on the message. If the message is to be Nack:ed I just don't call Ack and the server would resend the message after the timeout period (that could then be kept quite short).
Maybe something like a Ping() function on the message to tell the server that I am alive and working on a it.

**Comment by kozlovic** [comment]
@PatrLind As discussed [here](https://github.com/nats-io/stan.go/issues/167#issuecomment-481452373), have you thought about closing the subscription if you want the message to "nack'ed"? That is, as you do right now, set your AckWait to higher than max msg processing. If you experience a failure, close the subscription. Check on your error condition to know if it is resolved, if that is the case, restart subscription. Message will be redelivered.

**Comment by PatrLind** [comment]
@kozlovic, thank you, I did read that but I don't think I fully understood why that would help, but I think I get why now.
Maybe the request for Nack and similar functionality is a bit of an misunderstanding on how NATS Streaming works. It doesn't have things like delayed messages or dead letter queues and therefore things like this is not supposed to be handled inside NATS Streaming at all.
For my own system, I think I will add a better retry and failure logic at a higher up level, so that NATS only have to deal with client failures and not deal with processing failures and retries (other than suspending the queue temporarily for some specific short term failures).

**Comment by quintans** [comment]
@kozlovic I came to this thread because I was looking for an explicit `Nack` functionality. I am trying to implement a simple partitioning solution where I have two consumers. Each consumer will only process a message depending on a token inside the payload. When a message is not eligible to be processed by a consumer I wanted to `Nack` message in the hopes that it would be picked up by the other consumer. The solution discussed above, of closing the subscription, it is not a good fit for this.
The use of `Nack` here would be a good enough solution.
What would you propose instead of `Nack` or closing the subscription?

**Comment by froggyling** [comment]
I be thinking too simplistically here;

But from what I understand of the problem is that that if a message is can not be processed for any reason,. as per OP, if a message does not have some token needed for the consumer, they want to be able to have it pushed back for immediate process (but not to the consumer that sent it back).

From what I see proposed, closing a subscription would not be the best way of doing this, but indeed some sort of Nack call that asks for message to be resent (like a requeue) but with additional logic to not send to the consumer that Nack'd  

this could of course course a dead letter message if all consumers Nack. (technical resend but not to consumer that requested resend)

I can see a few uses for a Nack as described above, although you could also just have topics for each token, and consumers just process all messages broadcast out, and only Ack ones that match and lets NATS resend as needed (yes wasted cycles potentially getting same message sent back to the same queue/topic and the process delay of ack timeouts, etc...).

anyway just thinking aloud. But, feel a resend/ignore-me option (since above is not really a Nack, its a resend but not to me situation)

**Comment by quintans** [comment]
@froggyling I would say that it is an `Nack` at least for the consumer that picks it up. But ignoring the definition of `Nack`, what I am looking for is the `requeue` functionality that you describe.
Solving this challenge is possible, but not by creating topics since we would be talking about thousands if the token is an aggregate ID, but by forwarding the message to the correct node, using for example NATS request/reply and having only the appropriate node responding. But as I said in the beginning, the simplest approach would be if we had a Nack/Requeue out of the box.

**Comment by froggyling** [comment]
Agree on the topics if a lot of different tokens. (but even so scenario you describe implies 1-2 consumers per token <unless you are partitioning across a range of tokens that is) A basic requeue with flag to ignore the caller of the requeue would be interesting – although bears other costs (processing cycles/storage) If time permits, I will ponder this more and maybe see how much work it would be to do a simple proof of concept (although am a tad overloaded right now).   From: Paulo Quintans PereiraSent: Saturday, August 15, 2020 6:59 PMTo: nats-io/stan.goCc: froggyling; MentionSubject: Re: [nats-io/stan.go] Support for "NAck" (#167) @froggyling I would say that it is an Nack at least for the consumer that picks it up. But ignoring the definition of Nack, what I am looking for is the requeue functionality that you describe.Solving this challenge is possible, but not by creating topics since we would be talking about thousands if the token is an aggregate ID, but by forwarding the message to the correct node, using for example NATS request/reply and having only the appropriate node responding. But as I said in the beginning, the simplest approach would be if we had a Nack/Requeue out of the box.—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by quintans** [comment]
@froggyling thanks for thinking about it. I was talking about a simple partitioning using for example the mod of the hash of the ID.
In the consumer handler I would have:
```go
// consumer #0
if hash(ID)%2 != 0 {
  m.Requeue()  // <- nice to have
  return
}
```

**Comment by ripienaar** [comment]
I do not think NAK will come to Streaming Server. If you are only starting now you might be interested in JetStream - currently in preview - that has several features that can help with this including wildcard support and NAKs

https://github.com/nats-io/jetstream

**Comment by froggyling** [comment]
Oh I completely forgot about Jetstream (will have to see  how its progressed.)    From: R.I.PienaarSent: Saturday, August 15, 2020 7:46 PMTo: nats-io/stan.goCc: froggyling; MentionSubject: Re: [nats-io/stan.go] Support for "NAck" (#167) I do not think NAK will come to Streaming Server. If you are only starting now you might be interested in JetStream - currently in preview - that has several features that can help with this including wildcard support and NAKshttps://github.com/nats-io/jetstream—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by quintans** [comment]
@ripienaar Thanks. I will take a look.
@froggyling your replies are appearing with html tags https://github.com/nats-io/stan.go/issues/167#issuecomment-674442026

**Comment by froggyling** [comment]
Hmm that’s weird, thanks for letting me know

**Comment by kozlovic** [comment]
@quintans @froggyling As @ripienaar mentioned, you should be looking into JetStream, but I would also say that if your consumer wants to select which of the message to actually consume, it goes against the server design, which picks a consumer based on number of outstanding acks, responsiveness to heartbeat, and some randomness. I could argue that the method you use to decide how to consume: `hash(ID)%2` could be used by the producer and add that at the end of normal channel name. So you would still limit the number of channels to whatever the modulo you plan to use, but then as @froggyling mentioned, you would have simply consumer (as opposed to queue consumer) responsible to consume from each "partition".
Whatever approach, I think this goes against the idea of a queue consumer (at least in NATS) which is really meant to be used to distribute the load from a given "queue" to any consumer in that same group. If members of the same queue group are specialized and cannot/should not process some messages for this same group, then they should not be part of the group :-)

**Comment by quintans** [comment]
Thanks @kozlovic 
The functionality that I was really looking for was the one described [here](https://github.com/nats-io/jetstream/issues/103#issuecomment-674546316) and I was trying to see what would be the shortest (not the best) way of implementing this.
Of course this can be implemented as you say, but it is not an "elastic", meaning increasing the number of consumer would be a pain.
Anyway, I guess you are not adding new functionalities and all efforts are going into JetStream, and that makes sense.

**Comment by kozlovic** [comment]
Thank you for your understanding. Out of curiosity, which client library are you using (since if we added this feature it would have to be a client proto/lib change)?

**Comment by quintans** [comment]
I am using the nats stream client lib https://github.com/nats-io/stan.go

**Comment by kozlovic** [comment]
Thanks for letting me know. When I have time, I will see what it would take to implement this feature (but have to make sure it would work in clustering mode as well than standalone). But again, no promise especially that it requires client changes and then may lead to have to add that to several client libraries that we support, which again, is not the goal with JetStream being ready "soon".

**Comment by quintans** [comment]
Thanks, I really appreciate you taking the time to evaluate this.

**Comment by kozlovic** [comment]
I actually prototyped this feature and here are my thoughts:
- For non queue subscriptions, NAck(ignoreMe: bool) will cause message to be immediately redelivered.
  Specifying `ignoreMe: true` is an error for these subscriptions.
- For queue subscriptions:
 - NAck(ignoreMe: false) causes message to be immediately redelivered, possibly to the same queue member.
 - NAck(ignoreMe: true) causes message to be immediately redelivered to another queue member. If there is no other queue member, the message will be redelivered to the original subscription *after* the AckWait (to avoid possible rapid successions of NAck, redelivery, NAck, etc..)

Since for a given subscription the Ack/NAck need to be processed in the order they are sent (as for instance in the O/P use case of using a `defer msg.NAck()`), I made it so that the NAck protocol causes a unmarshal failure of the Ack and the server then tries to decode as a NAck. In normal Ack cases, this will have no impact, but if NAck are sent as often as Acks, I could see that this may not be optimal, although I don't think that the effect would be noticeable.

My concern is that if an application calls NAck and message is redelivered to it right away, this is going to create a "loop" and I think that we should have a circuit breaker such as say after N attempts to resend a NAck message to the same subscription - in a short amount of time? - then ignore the NAck request which will trigger a normal redelivery after the AckWait. That I have not implemented.

Thoughts?

**Comment by quintans** [comment]
@kozlovic I thought you were going to implement what I described here https://github.com/nats-io/jetstream/issues/103#issuecomment-674546316. This is what I was looking for.
Now I feel bad for for not being clear :(

**Comment by kozlovic** [comment]
@quintans What I prototyped is the NAck (this issue here with original poster) is about that, not about exclusive consumer. If subscribers to this issue want to comment about the NAck proposal, that is fine. I am not set in implementing anyway, wanted to have a feel if this would be possible or not.

**Comment by quintans** [comment]
Would you be open to explore the the key sharing feature if I created an issue for it?

**Comment by froggyling** [comment]
Sounds great, from my point of view

On the circuity breaker, tbh. In the case of re-delivery to same consumer if it’s the only one 
no matter what the Nack ignore me flag is -> responsibility of the consumer at that point.

That keeps flexibility in place, and simply pushing it back until next ackWait, is fine even if a bit noisy

As far as OP, should solve initial issue he wanted solved (without custom code for a specific case) 

For example, 

consumer -> process message -> no token in message that means I can keep processing to ack() -> send Nack(ignoreme)  

this way responsibility of logic used to do a Nack is purely down to desired Nack implementation (and message data payload)

ps. very much what I was trying to suggest earlier in this thread.

**Comment by froggyling** [comment]
In fact I have a couple use cases where it would save me some annoyances

right now, I sometimes need to push a message back into a Q for various reasons, the nack would save me doing an ack, and then producing a new message (lazy always wins for me)

**Comment by quintans** [comment]
Regarding the NAck, to avoid looping a backoff strategy could be defined. This could be done the consumer implementer if the attempt number is also delivered, or inside the client library.
